sklearn/linear_model/_glm/tests/test_glm.py

```diff
@@ -29,6 +47,272 @@ def regression_data():
+    """Test that GLM converges for all solvers to correct solution.
+
+    We work with a simple constructed data set with known solution.
+    """
+    model, X, y, _, coef_with_intercept, coef_without_intercept = glm_dataset
+    alpha = 1.0  # because glm_dataset uses this.
+    params = dict(
+        alpha=alpha,
+        fit_intercept=fit_intercept,
+        solver=solver,
+        tol=1e-12,
+        max_iter=1000,
+    )
+
+    model = clone(model).set_params(**params)
+    X = X[:, :-1]  # remove intercept
+    if fit_intercept:
+        coef = coef_with_intercept
+        intercept = coef[-1]
+        coef = coef[:-1]
+    else:
+        coef = coef_without_intercept
+        intercept = 0
+    model.fit(X, y)
+
+    rtol = 3e-5 if solver == "lbfgs" else 1e-11
+    assert model.intercept_ == pytest.approx(intercept, rel=rtol)
+    assert_allclose(model.coef_, coef, rtol=rtol)
+
+    # Same with sample_weight.
+    model = (
+        clone(model).set_params(**params).fit(X, y, sample_weight=np.ones(X.shape[0]))
+    )
+    assert model.intercept_ == pytest.approx(intercept, rel=rtol)
+    assert_allclose(model.coef_, coef, rtol=rtol)
+
+
+@pytest.mark.parametrize("solver", SOLVERS)
+@pytest.mark.parametrize("fit_intercept", [True, False])
+def test_glm_regression_hstacked_X(solver, fit_intercept, glm_dataset):
+    """Test that GLM converges for all solvers to correct solution on hstacked data.
+
+    We work with a simple constructed data set with known solution.
+    Fit on [X] with alpha is the same as fit on [X, X]/2 with alpha/2.
+    For long X, [X, X] is a singular matrix.
+    """
+    model, X, y, _, coef_with_intercept, coef_without_intercept = glm_dataset
+    n_samples, n_features = X.shape
+    alpha = 1.0  # because glm_dataset uses this.
+    params = dict(
+        alpha=alpha / 2,
+        fit_intercept=fit_intercept,
+        solver=solver,
+        tol=1e-12,
+        max_iter=1000,
+    )
+
+    if (
+        solver == "lbfgs"
+        and fit_intercept is False
+        and (
+            isinstance(model, BinomialRegressor)
+            or (isinstance(model, PoissonRegressor) and n_features > n_samples)
+        )
+    ):
+        # Line search cannot locate an adequate point after MAXLS
+        #  function and gradient evaluations.
+        #  Previous x, f and g restored.
+        # Possible causes: 1 error in function or gradient evaluation;
+        #                  2 rounding error dominate computation.
+        pytest.xfail()
+
+    model = clone(model).set_params(**params)
+    X = X[:, :-1]  # remove intercept
+    X = 0.5 * np.concatenate((X, X), axis=1)
+    assert np.linalg.matrix_rank(X) <= min(n_samples, n_features - 1)
+    if fit_intercept:
+        coef = coef_with_intercept
+        intercept = coef[-1]
+        coef = coef[:-1]
+    else:
+        coef = coef_without_intercept
+        intercept = 0
+    model.fit(X, y)
+
+    rtol = 3e-5 if solver == "lbfgs" else 1e-11
```

Comment:
At least, it's hard to achieve high precision with lbfgs.

Reply:
This is confirmed by https://github.com/scikit-learn/scikit-learn/pull/23314#issuecomment-1133642064 where I vary the `tol` parameter for all solvers.

It's even worse in our use of LBFGS in the `LogisticRegression`, probably because the `tol` criterion is not handled correctly but this is a different problem.